src/walberla_bridge/src/lattice_boltzmann/LBWalberlaImpl.hpp

```diff
@@ -389,8 +391,52 @@ class LBWalberlaImpl : public LBWalberlaBase {
     if (not lower_bc or not upper_bc) {
       return std::nullopt;
     }
-    assert(&(*(lower_bc->block)) == &(*(upper_bc->block)));
-    return {CellInterval(lower_bc->cell, upper_bc->cell)};
+    Cell const global_lower_cell = lower_bc->cell;
+    Cell const global_upper_cell =
+        Cell(static_cast<int>(upper_bc->cell[0] +
+                              upper_bc->block->getAABB().min()[0] -
+                              lower_bc->block->getAABB().min()[0]),
+             static_cast<int>(upper_bc->cell[1] +
+                              upper_bc->block->getAABB().min()[1] -
+                              lower_bc->block->getAABB().min()[1]),
+             static_cast<int>(upper_bc->cell[2] +
+                              upper_bc->block->getAABB().min()[2] -
+                              lower_bc->block->getAABB().min()[2]));
+    return {CellInterval(global_lower_cell, global_upper_cell)};
+  }
+
+  // Interval within local block
+  [[nodiscard]] std::optional<CellInterval> get_block_interval(
+      Utils::Vector3i const &lower_corner, Utils::Vector3i const &upper_corner,
+      Utils::Vector3i const &block_offset, IBlock const &block) const {
+    auto block_lower_corner = to_vector3i(block.getAABB().min());
+    if (upper_corner[0] < block_lower_corner[0] or
+        upper_corner[1] < block_lower_corner[1] or
+        upper_corner[2] < block_lower_corner[2]) {
+      return std::nullopt;
+    }
```

Comment:
In the past, ESPResSo was affected by subtle bugs where indices in the compared vectors wouldn't be the same on the left-hand-side and right-and-side of the comparison operator. To avoid this type of mistake, one can write here and elsewhere `if (not(block_lower_corner > upper_corner))`. Comparison operators on `Utils::Vector` objects will internally call `Utils::detail::all_of(lhs, rhs, cmp)`, which is why sometimes the lhs and rhs have to be exchanged.

Reply:
I have rewritten it, as you suggest.